### ISSUE TITLE
Delta backup results in signature mismatch 

### ISSUE THREAD

**Issue by koehn** [wal-g/wal-g#566]
My full backups work fine, and regular pushes of wal files work too. Thanks!

But a delta backup tends to fail with errors like this:

```
ERROR: 2020/02/18 03:10:11.086988 upload: could not Upload 'pg_wal/000000010000000B0000009F': failed to upload 'wal_005/000000010000000B0000009F.lz4' to bucket 'bucket': SignatureDoesNotMatch: The request signature we calculated does not match the signature you provided. Check your key and signing method.
```

I’m using Minio as my S3 server. 

Could this happen because it takes too long to compute the delta file, and the time-based signature is computed too soon, and thus expires?

**Comment by x4m** [comment]
Seems like you are uploading wal, not backup
to enable delta backups you need to change WALG_DELTA_MAX_STEPS to nonzero value

**Comment by koehn** [comment]
Yes, I’ve done that. This is an error I see while the delta backup is running.

**Comment by koehn** [comment]
Since I’ve just run into this again (four years later), here’s a working NGINX configuration:
```
server {

    server_name s3.example.com;

    ignore_invalid_headers off;
    client_max_body_size 0;
    client_body_buffer_size 8m;
    proxy_buffering off;

    location / {
        proxy_pass http://10.0.1.123:1234/;
        proxy_buffering off;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header Host $http_host;

        proxy_connect_timeout  300;
        # Default is HTTP/1, keepalive is only enabled in HTTP/1.1
        proxy_http_version 1.1;
        proxy_set_header Connection "";

    }
    … yadda yadda
}
```

**Comment by karakumium** [comment]
hi @koehn!

Seems like I have the same problem, but probably with a different infrastructure configuration.

Is it possible for you to provide a diff between your previous (where you got SignatureDoesNotMatch error) and working Nginx configuration?
> here’s a working NGINX configuration



Thank you in advance

**Comment by chipitsine** [comment]
nginx is known to break S3 signatures, because it violates http rfc by not passing "Expect" header, which is included to S3 signature.
it is known for signature v4.

please either switch to v2 (which is not affected by Expect header drop) or switch to stream proxy (instead of http).

anyway, it is not a wal-g issue

**Comment by koehn** [comment]
Hey all, I’ve since switched from the soon-to-be-unsupported ingress-nginx to Traefik, so this is a non-issue for me.

**Comment by ostinru** [comment]
We have fixed `SignatureDoesNotMatch` by setting `S3_DISABLE_100_CONTINUE: true`. #2042

**Comment by chipitsine** [comment]
Expect is end to end header and dropping it on some hop really violates RFC.

but disabling Expect in favour of nginx behaviour is smart :)
